File: apps/consumer/ios/WatchApp/WatchHandoffLink.swift
import Foundation

enum WatchHandoffLink {
  static func buildLink(for order: WatchReorder) -> String {
    if let fuel = order.fuel {
      return buildFuelLink(storeId: order.storeId, fuel: fuel, currency: order.currency ?? "")
    }
    return buildReorderLink(order: order)
  }

  static func buildOrderPaymentLink(orderId: String, payment: WatchOffSessionResult) -> String? {
    guard let clientSecret = payment.clientSecret, !clientSecret.isEmpty else { return nil }
    var parts = [
      "orderId=\(encode(orderId))",
      "clientSecret=\(encode(clientSecret))"
    ]
    if let paymentId = payment.paymentId, !paymentId.isEmpty {
      parts.append("paymentId=\(encode(paymentId))")
    }
    if let intentId = payment.paymentIntentId, !intentId.isEmpty {
      parts.append("paymentIntentId=\(encode(intentId))")
    }
    if let customerId = payment.customerId, !customerId.isEmpty {
      parts.append("customerId=\(encode(customerId))")
    }
    if let ephemeralKey = payment.ephemeralKey, !ephemeralKey.isEmpty {
      parts.append("ephemeralKey=\(encode(ephemeralKey))")
    }
    if let accountId = payment.stripeAccountId, !accountId.isEmpty {
      parts.append("stripeAccountId=\(encode(accountId))")
    }
    if let publishableKey = payment.publishableKey, !publishableKey.isEmpty {
      parts.append("publishableKey=\(encode(publishableKey))")
    }
    if let amount = payment.amountCents, amount > 0 {
      parts.append("amountCents=\(amount)")
    }
    if let currency = payment.currency, !currency.isEmpty {
      parts.append("currency=\(encode(currency))")
    }
    return "com.orderingintelligence.consumer://order-payment?\(parts.joined(separator: "&"))"
  }

  private static func buildReorderLink(order: WatchReorder) -> String {
    let items = order.items.map { "\(encode($0.itemId)):\($0.quantity)" }.joined(separator: ",")
    var parts = [
      "storeId=\(encode(order.storeId))",
      "items=\(items)",
    ]
    if !order.title.isEmpty {
      parts.append("title=\(encode(order.title))")
    }
    if let currency = order.currency, !currency.isEmpty {
      parts.append("currency=\(encode(currency))")
    }
    return "com.orderingintelligence.consumer://reorder?\(parts.joined(separator: "&"))"
  }

  private static func buildFuelLink(storeId: String, fuel: WatchFuelOrder, currency: String) -> String {
    var parts: [String] = [
      "storeId=\(encode(storeId))",
      "gradeId=\(encode(fuel.fuelGradeId))",
    ]
    if let name = fuel.fuelGradeName, !name.isEmpty {
      parts.append("gradeName=\(encode(name))")
    }
    if let price = fuel.unitPriceCents, price > 0 {
      parts.append("unitPriceCents=\(price)")
    }
    let flow = fuel.paymentFlow ?? "prepay"
    parts.append("paymentFlow=\(encode(flow))")
    if let amount = fuel.requestedAmountCents, amount > 0 {
      parts.append("requestedAmountCents=\(amount)")
    }
    if let liters = fuel.requestedLiters, liters > 0 {
      parts.append("requestedLiters=\(liters)")
    }
    if let preauth = fuel.preauthAmountCents, preauth > 0 {
      parts.append("preauthAmountCents=\(preauth)")
    }
    if !currency.isEmpty {
      parts.append("currency=\(encode(currency))")
    }
    return "com.orderingintelligence.consumer://fuel-order?\(parts.joined(separator: "&"))"
  }

  private static func encode(_ value: String) -> String {
    return value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value
  }
}

File: apps/consumer/android/wear/src/main/java/com/orderingintelligence/consumer/wear/WatchHandoffLink.kt
package com.orderingintelligence.consumer.wear

import java.net.URLEncoder

object WatchHandoffLink {
  fun buildLink(order: WatchReorder): String {
    val fuel = order.fuel
    if (fuel != null) {
      return buildFuelLink(order.storeId, fuel, order.currency)
    }
    return buildReorderLink(order)
  }

  fun buildOrderPaymentLink(orderId: String, payment: WatchOffSessionResult): String? {
    val clientSecret = payment.clientSecret ?: return null
    val params = mutableListOf(
      "orderId=${encode(orderId)}",
      "clientSecret=${encode(clientSecret)}"
    )
    payment.paymentId?.let { if (it.isNotBlank()) params.add("paymentId=${encode(it)}") }
    payment.paymentIntentId?.let { if (it.isNotBlank()) params.add("paymentIntentId=${encode(it)}") }
    payment.customerId?.let { if (it.isNotBlank()) params.add("customerId=${encode(it)}") }
    payment.ephemeralKey?.let { if (it.isNotBlank()) params.add("ephemeralKey=${encode(it)}") }
    payment.stripeAccountId?.let { if (it.isNotBlank()) params.add("stripeAccountId=${encode(it)}") }
    payment.publishableKey?.let { if (it.isNotBlank()) params.add("publishableKey=${encode(it)}") }
    payment.amountCents?.let { if (it > 0) params.add("amountCents=$it") }
    payment.currency?.let { if (it.isNotBlank()) params.add("currency=${encode(it)}") }
    return "com.orderingintelligence.consumer://order-payment?${params.joinToString("&")}"
  }

  private fun buildReorderLink(order: WatchReorder): String {
    val itemsParam = order.items.joinToString(",") { item ->
      "${encode(item.itemId)}:${item.quantity}"
    }
    val params = mutableListOf(
      "storeId=${encode(order.storeId)}",
      "items=$itemsParam"
    )
    if (order.title.isNotBlank()) params.add("title=${encode(order.title)}")
    if (order.currency.isNotBlank()) params.add("currency=${encode(order.currency)}")
    return "com.orderingintelligence.consumer://reorder?${params.joinToString("&")}"
  }

  private fun buildFuelLink(storeId: String, fuel: WatchFuelOrder, currency: String): String {
    val params = mutableListOf(
      "storeId=${encode(storeId)}",
      "gradeId=${encode(fuel.fuelGradeId)}"
    )
    if (fuel.fuelGradeName.isNotBlank()) {
      params.add("gradeName=${encode(fuel.fuelGradeName)}")
    }
    if (fuel.unitPriceCents > 0) params.add("unitPriceCents=${fuel.unitPriceCents}")
    val flow = fuel.paymentFlow.ifBlank { "prepay" }
    params.add("paymentFlow=${encode(flow)}")
    if (fuel.requestedAmountCents > 0) params.add("requestedAmountCents=${fuel.requestedAmountCents}")
    if (fuel.requestedLiters > 0) params.add("requestedLiters=${fuel.requestedLiters}")
    if (fuel.preauthAmountCents > 0) params.add("preauthAmountCents=${fuel.preauthAmountCents}")
    if (currency.isNotBlank()) params.add("currency=${encode(currency)}")
    return "com.orderingintelligence.consumer://fuel-order?${params.joinToString("&")}"
  }

  private fun encode(value: String): String = URLEncoder.encode(value, "UTF-8")
}
